messages/compiler/cpp/cpp_visitor.py

```diff
@@ -0,0 +1,261 @@
+    {name} value;
+    deserialize(start, end, value);
+    val = value;
+    return;
+  }}
+"""
+    s += """
+  throw cmf::DeserializeError(std::string("Invalid Message id in variant: ") + std::to_string(id));
+}
+"""
+    return s
+
+
+def equalop_str(msg_name, fields):
+    """ Create an 'operator==' function for the current message struct """
+    comparison = " && ".join([f"l.{f} == r.{f}" for f in fields])
+    return f"""\
+bool operator==(const {msg_name}& l, const {msg_name}& r) {{
+  return {comparison};
+}}"""
+
+
+class CppVisitor(Visitor):
+    """ A visitor that generates C++ code. """
+    def __init__(self):
+        # All output currently constructed
+        self.output = ""
+
+        # For each message, all four of the following C++ code strings are constructed in the same
+        # pass.
+
+        # The current message being processed
+        self.msg_name = ''
+
+        # The current field being processed for a message
+        self.field = {'type': '', 'name': ''}
+
+        # All fields currently seen for the given message
+        self.fields_seen = []
+
+        # The struct being created for the current message. This includes the fields of the struct.
+        self.struct = ""
+
+        # The 'serialize' function for the current message
+        self.serialize = ""
+
+        # The 'deserialize' function for the current message
+        self.deserialize = ""
+
+        # Each oneof in a message corresponds to a variant. Since we don't need duplicate
+        # serialization functions, in case there are multiple messages or fields with the same
+        # variants, we only generate a single serialization and deserialization function for each
+        # variant.
+        #
+        # This set keeps track of which variants already had their serialization and
+        # deserialization functions generated.
+        self.oneofs_seen = set()
+
+        # The `serialize` member functions for all oneofs in the current message
+        self.oneof_serialize = ""
+
+        # The `deserialize` member functions for all oneofs in the current message
+        self.oneof_deserialize = ""
+
+    def msg_start(self, name, id):
+        self.msg_name = name
+        self.struct = struct_start(name, id)
+        self.serialize = serialize_start(name)
+        self.deserialize = deserialize_start(name)
+
+    def msg_end(self):
+        self.struct += "};\n"
+        self.serialize += "}"
+        self.deserialize += deserialize_end(self.msg_name)
+        self.output += "\n".join([
+            s for s in [
+                self.struct,
+                self.oneof_serialize,
+                self.oneof_deserialize,
+                equalop_str(self.msg_name, self.fields_seen),
+                self.serialize,
+                self.deserialize,
+            ] if s != ''
+        ])
+        # output and oneofs_seen accumulate across messages
+        output = self.output
+        oneofs = self.oneofs_seen
+        self.__init__()
```

Comment:
Can we call it `reset` or `clean`?

Reply:
__init__() is the "magic" constructor function. I just reused it rather than adding another method.

Reply:
I know what `__init__` is. I was not sure if the intention was clear.
Nevermind, I do have a strong opinion on this.

Reply:
Thinking about this, I think you are right. It will be clearer to the reader with an alias method. I will add one. Thanks!